### PR TITLE
DEP: data_dirs

### PR DESCRIPTION
**Pull request by rstoneback**
# Description

Addresses # 681, 683

Provides Deprecation notices for use of `pysat.data_dir` and `pysat.utils.set_data_dir`. Getting a warning to pop up on a simple module variable is not the easiest thing to do. Turns out I had to wrap the python module in a class and then get that class to pretend it is a module. The core Python developers actually put in time to make this work because they liked it.

https://stackoverflow.com/questions/2447353/getattr-on-a-module/7668273#7668273

Tests pass on my local machine but aren't quite passing on TravisCI yet.

```
ERROR: Test basic seasonal 1D median
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/pysat/pysat/pysat/tests/test_ssnl_ave.py", line 344, in test_basic_seasonal_median1D
    [self.dname])
  File "/home/travis/build/pysat/pysat/pysat/ssnl/avg.py", line 70, in median1D
    raise ValueError(estr)
ValueError: Parameter must be an Instrument or a Constellation. Received: <class 'pysat._instrument.Instrument'>
```
I'm not exactly sure why but it has to be some detail about the type that is popping out from the whole class wrapped module thing.

On the plus side though,
```
In [1]: import pysat

In [2]: pysat.data_dir
/Users/russellstoneback/opt/miniconda3/envs/standard2/bin/ipython:1: DeprecationWarning: `pysat.data_dir` has been deprecated in pysat 3.0.0 and replaced with `pysat.params['data_dirs']`.
  #!/Users/russellstoneback/opt/miniconda3/envs/standard2/bin/python

``` 

## Type of change

Please delete options that are not relevant.

- Deprecation Update

# How Has This Been Tested?

local unit tests (passing) and TravisCI (not currently passing)

# Checklist:

- [x] Make sure you are merging into the ``develop`` (not ``main``) branch
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] Add a note to ``CHANGELOG.md``, summarizing the changes

### CODE REVIEW

**Comment by rstoneback** [comment]
@aburrell and @jklenzing this code is not ready for full review but adding a deprecation warning to `pysat.data_dir` requires a larger change than other parameters. How do y'all feel about wrapping pysat in a class that acts like a module?

File: pysat/__init__.py [hunk 1]
```diff
@@ -68,30 +68,154 @@
...
+if sys.version_info[0] >= 3:
+    from importlib import reload as re_load
+else:
+    re_load = reload
+
+
+def pysat_reload():
+    """Enable most of the underlying functionality of reload.
+
+    To provide a deprecation notice for `pysat.data_dir` the pysat module
```
Comment (aburrell) [inline pysat/__init__.py:101]:
```suggestion

    Note
    -----
    To provide a deprecation notice for `pysat.data_dir` the pysat module
```

File: pysat/ssnl/avg.py [hunk 2]
```diff
@@ -65,7 +65,9 @@ def median1D(const, bin1, label1, data_label, auto_bin=True, returnData=False):
     if isinstance(const, pysat.Instrument):
         const = [const]
     elif not isinstance(const, pysat.Constellation):
```
Comment (aburrell) [inline pysat/ssnl/avg.py:67]: Recommend adding a second evaluation for `pysat._instrument.Instrument`, etc.

File: pysat/tests/test_utils.py [hunk 3]
```diff
@@ -56,10 +56,58 @@ def test_deprecation_warning_computational_form():
...
+            pysat.utils.set_data_dir('.')
+
+        found_msg = False
+        for iwar in war:
+            print(iwar.message)
+            if iwar.category == DeprecationWarning \
+                    and str(iwar.message).find(wmsg) >= 0:
+                found_msg = True
+
+        assert found_msg, "didn't find warning about: {:}".format(wmsg)
```
Comment (aburrell) [inline pysat/tests/test_utils.py:85]: Recommend using warning evaluation function

File: pysat/tests/test_utils.py [hunk 4]
```diff
@@ -71,35 +119,49 @@ def teardown(self):
     # test pysat data dir options
     def test_set_data_dir(self):
         """update data_dir"""
+        import pysat
```
Comment (aburrell) [inline pysat/tests/test_utils.py:122]: Is it really correct to double import?  Would reload be more appropriate?

File: .travis.yml [hunk 5]
```diff
@@ -60,7 +60,7 @@ before_script:
 
 # command to run tests
 script:
- - nosetests -vs --with-coverage --cover-package=pysat
+ - nosetests -vs --with-coverage --cover-package=pysat.tests
```
Comment (aburrell) [inline .travis.yml:63]: Why is this changed?

**Review by aburrell — COMMENTED** [review]
Comments, suggestions, and questions

**Comment by aburrell** [comment]
> How do y'all feel about wrapping pysat in a class that acts like a module?

It seemed to solve the problem, so I have no negative feelings 🐍

**Comment by aburrell** [comment]
> Addresses # (issue)

Don't forget to link issues.

File: .travis.yml (hunk 5 above)
Comment (rstoneback) [reply .travis.yml:63]: It didn't work otherwise. I was doing most of the testing with `nosetests pysat.tests.whatever` so I didn't catch it at first. I added an import for tests in init and that didn't do the trick either. Something related to the wrapper class I presume.

File: pysat/tests/test_utils.py (hunk 4 above)
Comment (rstoneback) [reply pysat/tests/test_utils.py:122]: reload is bit weird now. reload doesn't work directly because it complains pysat isn't a module. `type(pysat)` returns different information now. I had to do a workaround to get the effect of a reload, which is the `pysat.pysat_reload()` but then you have to pull in the updated module from the global namespace, `pysat = sys.modules['pysat']`. Once there is a line `pysat = something` in the code the Python interpreter starts to complain if pysat isn't explicitly defined in more places.

File: pysat/tests/test_utils.py (hunk 3 above)
Comment (rstoneback) [reply pysat/tests/test_utils.py:85]: Happy to do that but I don't know what you mean. I grabbed this code from somewhere else in pysat. I'll look into it.

File: pysat/ssnl/avg.py (hunk 2 above)
Comment (rstoneback) [reply pysat/ssnl/avg.py:67]: Will do. As far as I know that string is correct. If I check things locally, 
```
In [7]: test = pysat.Instrument('pysat', 'testing',
   ...:                         clean_level='clean',
   ...:                         orbit_info={'index': 'mlt'},
   ...:                         update_files=True)
/Users/russellstoneback/opt/miniconda3/envs/standard2/bin/ipython:4: DeprecationWarning: Instrument attributes [units_label, name_label, notes_label, desc_label, min_label, max_label, fill_label] will be accessible through `Instrument.meta.labels` in pysat 3.0.0
  import re
/Users/russellstoneback/opt/miniconda3/envs/standard2/bin/ipython:4: DeprecationWarning: Meta labels [plot_label, axis_label, scale_label] are no longer standard metadata quantities in pysat 3.0.0
  import re
/Users/russellstoneback/opt/miniconda3/envs/standard2/bin/ipython:4: DeprecationWarning: Strict times will eventually be enforced upon all instruments. (strict_time_flag)
  import re

In [8]: type(test)
Out[8]: pysat._instrument.Instrument

In [9]: isinstance(test, pysat.Instrument)
Out[9]: True
```

It works out. The line in ssnl that is causing the problem is:

```
    if isinstance(const, pysat.Instrument):
        const = [const]
    elif not isinstance(const, pysat.Constellation):
        estr = ''.join(("Parameter must be an Instrument or a Constellation.",
                        " Received: ", str(type(const))))
        raise ValueError(estr)
```
the `isinstance(const, pysat.Instrument)` should evaluate to True like it does in In[9] above. I think this may actually be a weird side effect of reloading pysat in the new class wrapped module regime. It is possible the `pysat` that was originally attached to the ssnl module when it was first imported is now a different instance of `pysat` in the namespace that created the test instrument supplied to the method..... I can probably test that.

File: pysat/tests/test_utils.py (hunk 3 above)
Comment (rstoneback) [reply pysat/tests/test_utils.py:85]: The function in #755?

Comment (aburrell) [reply pysat/tests/test_utils.py:85]: Yup!  There are a few examples in the unit tests.  Now that you have the function name, I'm sure it's easy enough to find.

**Review by aburrell — APPROVED** [review]
